### PR TITLE
Ray namespaces

### PR DESCRIPTION
**Pull request by wuisawesome**
## Why are these changes needed?

This PR implements the namespaces feature described in #15684. It is currently only accessible via the private API `ray.init(job_config=JobConfig(namespace="...")`, because the client/session builder api isn't ready yet.

See the docs for more details. https://ray--15774.org.readthedocs.build/en/15774/namespaces.html

## Related issue number

#15684 

#15557 

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by wuisawesome** [comment]
@edoakes here's a draft of implementing namespaces mostly relying on the job config. IMO if we consider namespaces to be a property of a job (and not just an actor) this is the "right" way to do it (though I agree we could clean up the job config code...) 

If you agree that this is looks like a good start, I can write some tests and we can get this merged. If not, we can just use it for the demo then throw it out.

**Comment by WangTaoTheTonic** [comment]
Hi Alex, this is a nice feature and we have already have an internal version about this.
Do you have a draft design document about this? We can take a look and discuss together.
//cc @raulchen @wumuzi520

**Comment by wumuzi520** [comment]
> Hi Alex, this is a nice feature and we have already have an internal version about this.
> Do you have a draft design document about this? We can take a look and discuss together.
> //cc @raulchen @wumuzi520

@wuisawesome  @WangTaoTheTonic The namespace feature has been running internally for a year. We use the it to isolate different jobs.

File: src/ray/gcs/gcs_server/gcs_actor_manager.cc [hunk 1]
```diff
@@ -72,6 +72,8 @@ bool GcsActor::IsDetached() const { return actor_table_data_.is_detached(); }
 
 std::string GcsActor::GetName() const { return actor_table_data_.name(); }
 
+std::string GcsActor::GetNamespace() const { return actor_table_data_.ray_namespace(); }
```
Comment (edoakes) [inline src/ray/gcs/gcs_server/gcs_actor_manager.cc:75]: should be consistent in naming -- either just call it "namespace" or "ray namespace" everywhere. Preference for the former.

File: src/ray/gcs/gcs_server/gcs_actor_manager.cc [hunk 2]
```diff
@@ -708,9 +728,17 @@ void GcsActorManager::ReconstructActor(
...
-      if (it != named_actors_.end() && it->second == actor->GetActorID()) {
-        named_actors_.erase(it);
+      auto namespace_it = named_actors_.find(actor->GetNamespace());
+      if (namespace_it != named_actors_.end()) {
+        auto it = namespace_it->second.find(actor->GetName());
+        if (it != namespace_it->second.end()) {
+          namespace_it->second.erase(it);
+        }
+        // If we just removed the last actor in the namespace, remove don't leak the empty
+        // map.
```
Comment (edoakes) [inline src/ray/gcs/gcs_server/gcs_actor_manager.cc:738]:
```suggestion
        // If we just removed the last actor in the namespace, remove the empty map.
```

**Review by edoakes — COMMENTED** [review]
Looks pretty straightforward. I would prefer to not stuff this in the `JobConfig` and instead pass it directly in the task spec though (we are likely going to remove the `JobConfig` in the near future because it essentially only has runtime_env fields right now)

**Comment by edoakes** [comment]
@WangTaoTheTonic @wumuzi520 I just made an [RFC](https://github.com/ray-project/ray/issues/15557) for this, please take a look and leave your thoughts there! :)

File: src/ray/gcs/gcs_server/gcs_actor_manager.h [hunk 3]
```diff
@@ -441,6 +449,9 @@ class GcsActorManager : public rpc::ActorInfoHandler {
   /// This method MUST BE IDEMPOTENT because it can be called multiple times during
   /// actor destroy process.
   std::function<void(const ActorID &)> destroy_owned_placement_group_if_needed_;
+  /// A callback to get the namespace an actor belongs to based on its job id. This is
+  /// necessary for actor creation.
+  std::function<std::string(const JobID &)> get_namespace_;
```
Comment (wuisawesome) [inline src/ray/gcs/gcs_server/gcs_actor_manager.h:454]: TODO We need to populate this mapping on initialization in case of a GCS restart.

**Comment by wuisawesome** [comment]
> I would prefer to not stuff this in the JobConfig and instead pass it directly in the task spec though

I see where you're coming from, but we also have a requirement that we have to be able to associate a namespace with a job, even if there are no actors (like in a task only workload). 

How would we get that information? We could try to associate the namespace with the driver in the ray client server, but it feels like GCS should track this as a property of the job.

File: src/ray/gcs/gcs_server/gcs_actor_manager.cc (hunk 1 above)
Comment (wuisawesome) [reply src/ray/gcs/gcs_server/gcs_actor_manager.cc:75]: I have a preference for `namespace` too, but it's a reserved keyword in C++.

**Comment by edoakes** [comment]
@wuisawesome if we need to associate it at the job level that seems fine to me

**Comment by wuisawesome** [comment]
Closing in favor of #15774

**Pull request closed without merging**